net/peer_tracker.go

```diff
@@ -0,0 +1,68 @@
+package net
+
+import (
+	"sync"
+
+	logging "github.com/ipfs/go-log"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+var log = logging.Logger("peer-tracker")
+
+// PeerTracker is used to record a subset of peers. Its methods are thread safe.
+// It is designed to plug directly into libp2p disconnect notifications to
+// automatically register dropped connections.
+type PeerTracker struct {
+	// mu protects peers
+	mu sync.RWMutex
+	// peers maps stringified peer.IDs to info about their chains
+	peers map[string]struct{}
+}
+
+// NewPeerTracker creates a peer tracker.
+func NewPeerTracker() *PeerTracker {
+	return &PeerTracker{
+		peers: make(map[string]struct{}),
+	}
+}
+
+// Track adds information about a given peer.ID
+func (tracker *PeerTracker) Track(pid peer.ID) {
+	tracker.mu.Lock()
+	defer tracker.mu.Unlock()
+
+	pidKey := peer.IDB58Encode(pid)
+	if _, tracking := tracker.peers[pidKey]; tracking {
+		log.Warningf("unexpected duplicate track on peer: %s", pid)
```

Comment:
A duplicate hello without a disconnection shouldn't happen so I don't want to try to handle it like it is an expected thing.

In the unlikely event this does end up happening we should assume the remote node is reporting its current head in the latest invocation and not try to keep around stale state.

Reply:
^ Contradicting myself a little.  Perhaps we should error explicitly if its a dup?  Waddayou think?

Reply:
> In the unlikely event this does end up happening we should assume the remote node is reporting its current head in the latest invocation and not try to keep around stale state.

Agreed, I have redacted my previous comment.

Reply:
Although if the PeerTracker is used to track peers it hears about outside the hello protocol this may need rethinking -- hard to tell if that will be the case now though. Looks good.